app/services/move_work_package_service.rb

```diff
@@ -0,0 +1,191 @@
+    else
+      WorkPackage.transaction do
+        move_without_transaction(new_project, new_type, options) ||
+          raise(ActiveRecord::Rollback)
+      end || false
+    end
+  end
+
+  private
+
+  def move_without_transaction(new_project, new_type = nil, options = {})
+    attributes = options[:attributes] || {}
+
+    modified_work_package = copy_or_move(options[:copy], new_project, new_type, attributes)
+
+    if options[:copy]
+      return false unless copy(modified_work_package, attributes, options)
+    else
+      return false unless move(modified_work_package, new_project, options)
+    end
+
+    modified_work_package
+  end
+
+  def copy_or_move(make_copy, new_project, new_type, attributes)
+    modified_work_package = if make_copy
+                              WorkPackage.new.copy_from(work_package)
+                            else
+                              work_package
+                            end
+
+    move_to_project(modified_work_package, new_project)
+
+    move_to_type(modified_work_package, new_type)
+
+    bulk_assign_attributes(modified_work_package, attributes)
+
+    modified_work_package
+  end
+
+  def copy(modified_work_package, attributes, options)
+    set_default_values_on_copy(modified_work_package, attributes)
+
+    return false unless modified_work_package.save
+
+    create_and_save_journal_note modified_work_package, options[:journal_note]
+
+    true
+  end
+
+  def move(modified_work_package, new_project, options)
+    if options[:journal_note]
+      modified_work_package.add_journal user, options[:journal_note]
+    end
+
+    return false unless modified_work_package.save
+
+    move_time_entries(modified_work_package, new_project)
+
+    return false unless move_children(modified_work_package, new_project, options)
+
+    true
+  end
+
+  def move_to_project(work_package, new_project)
+    if new_project &&
+       work_package.project_id != new_project.id &&
+       allowed_to_move_to_project?(new_project)
+
+      work_package.delete_relations(work_package)
+
+      reassign_category(work_package, new_project)
+
+      # Keep the fixed_version if it's still valid in the new_project
+      unless new_project.shared_versions.include?(work_package.fixed_version)
+        work_package.fixed_version = nil
+      end
+
+      work_package.project = new_project
+
+      enforce_cross_project_settings(work_package)
+    end
+  end
+
+  def move_to_type(work_package, new_type)
+    if new_type
+      work_package.type = new_type
+      work_package.reset_custom_values!
+    end
+  end
+
+  def bulk_assign_attributes(work_package, attributes)
+    # Allow bulk setting of attributes on the work_package
+    if attributes
+      # before setting the attributes, we need to remove the move-related fields
+      work_package.attributes =
+        attributes.except(:copy, :new_project_id, :new_type_id, :follow, :ids)
+          .reject { |_key, value| value.blank? }
```

Comment:
I know that this is older than your PR. Do you know the intention of this `reject`?

I am asking because it will happily reject `false` values. In the case of custom fields such code already caused trouble, I don't think that we have any non-custom boolean attributes, but I better ask twice ^^